### DISCUSSION TITLE
NiceGUI uses 100% of a CPU core when running inside a container

### DISCUSSION THREAD

**Discussion by pwt** [zauberzeug/nicegui#2998] [Q&A]
### Question

Hi folks:

I'm making productive use of NiceGUI and finding it very elegant to use. However, I've run into an issue when trying to deploy an app within a container.

I'm sure I'm doing something dumb here, and I'm hoping someone can point out what. Ideas welcome!

# Problem

When I start a NiceGUI app within a container, it will use 100% of a CPU core. I've simplified my setup as far as I can in an attempt to narrow down what the issue could be, as follows:

# Dockerfile

```Dockerfile
FROM python:3.12-slim
RUN pip install --root-user-action=ignore --disable-pip-version-check --no-cache-dir nicegui
COPY nicegui-test.py nicegui-test.py
CMD python nicegui-test.py
```

# NiceGUI App: `nicegui-test.py`

```python
from nicegui import ui
ui.label('Hello NiceGUI!')
ui.run()
```

# Build the Container Image

```shell
docker build -t nicegui-test .
```

# Run the Container

```shell
% docker run --rm -p 8080:8080 --name nicegui-run nicegui-test
NiceGUI ready to go on http://localhost:8080, and http://172.17.0.2:8080
```

This works, but uses a continuous 100% of a CPU core.

# Run the NiceGUI App from Bash within a Container

I see the same issue if I start the NiceGUI app from a Bash shell within the container:

```shell
% docker run -it --rm -p 8080:8080 --name nicegui-run nicegui-test bash
root@d589ccb0d5d3:/# python nicegui-test.py
NiceGUI ready to go on http://localhost:8080, and http://172.17.0.2:8080
```

The CPU will latch at 100% once the app is started.

Thanks for any pointers, folks.

# Stack Trace

**Edit**: added stack trace when the app is interrupted:

```text
NiceGUI ready to go on http://localhost:8080, and http://172.17.0.2:8080
^CProcess SpawnProcess-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.12/site-packages/uvicorn/_subprocess.py", line 78, in subprocess_started
    target(sockets=sockets)
  File "/usr/local/lib/python3.12/site-packages/nicegui/server.py", line 36, in run
    super().run(sockets=sockets)
  File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 65, in run
    return asyncio.run(self.serve(sockets=sockets))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 123, in run
    raise KeyboardInterrupt()
KeyboardInterrupt
```

**Comment by rodja** [comment]
I briefly looked at it and could reproduce the behaviour. Strangely when I use the `zauberzeug/nicegui` docker image the cpu utilization was ok. Has anyone an idea what is going on?

**Comment by rodja** [comment]
I transferred it into an issue: https://github.com/zauberzeug/nicegui/issues/3066

**Comment by pwt** [comment]
Thanks for converting this into an issue.

It looks like using `ui.run(reload=False)` fixes the issue, which is actually fine for my purposes.

**Comment by thiagomoretto** [comment]
Do you have a venv or deps in the same path level as the app?

I had similar issue but it was related to uvicorn, not specific to Nicegui or even Fastapi

Anyway, If someone steps by here what I have found is that using just `--reload` and on the same path having a venv or hundreds of python packages (the deps), uvicorn would end up watching for modification of these thousands of files. The alternative is to provide a directory to watch using `--reload-dir [your-source-path]` to narrow its scope.

As you have mentioned that you're using ui.run instead, if that's the same case you can use `uvicorn_reload_dirs` param on `ui.run`.

**Reply by falkoschindler** [reply]
Thanks, @thiagomoretto! You pointed me into the right direction. Indeed the watchdog seems to be the problem when running in `/`. See https://github.com/zauberzeug/nicegui/issues/3066#issuecomment-2142468480.
